File: Assn5.cpp
#include <iostream>
#include <string> 
#include <sstream>
using namespace std;

// function declaration
void build(string, int, int, string, int, string, int, int);

int main(int argc, char *argv[]) {
  string number = argv[1];
  int sum = stoi(argv[2]);
  int intNumberLength = number.length();
  if (intNumberLength == 1) // prevents going into recursive loop
  {
    if (stoi(number) == sum)
    {
      cout << number << "=" << sum;
    }
  } else { // enter recursive loop - once with original string, another with negative string
      build(number, 1, number.at(0) - '0', string(1, number.at(0)), number.at(0) - '0', "", sum, intNumberLength);
      build("-" + number, 2, (number.at(0) - '0') * -1, "-" + string(1, number.at(0)), (number.at(0) - '0') * -1, "", sum, intNumberLength + 1);
  }
  return 0;
} // end main

/*
   * Create and evaluate expressions through recursive method.
   * number = original string
   * index = current index in number string
   * runningSum = evaluted sum of current expression
   * expression = string representation of current combination of operators+operands
   * prev = most recent operand
   * prevop = previous operator (+, -, or _/concatenation)
   * sum = target sum for comparison
   * length = length of number string
*/
void build(string number, int index, int runningSum, string expression, int prev, string prevop, int sum, int length) {
  // base case
  if (index == length - 1)
  {
    // evaluate + last operand
    if (runningSum + (number.at(index) - '0') == sum)
    {
      cout << expression << "+" << number.at(index) << "=" << sum << "\n";
    }
    // evaluate - last operand
    if (runningSum - (number.at(index) - '0') == sum)
    {
      cout << expression << "-" << number.at(index) << "=" << sum << "\n";
    }
    // evaluate operand concatenated with previous operand
    int newEnd = stoi(to_string(prev) + number.at(index)); // concatenate previous operand with final digit
    if (prevop.compare("+") == 0 && ((runningSum - prev + newEnd) == sum))
    {
      cout << expression << number.at(index) << "=" << sum << "\n";
    } // if prevop = +, subtract previous operand and add new final operand
    else if (prevop.compare("-") == 0 && ((runningSum + prev - newEnd) == sum))
    {
      cout << expression << number.at(index) << "=" << sum << "\n";
    } // if prevop = -, add previous operand and subtract new final operand
    else if (prevop.compare("") == 0)
    {
      expression += number.at(index);
      if (stoi(expression) == sum) // evaluate entire expression
      {
        cout << expression << "=" << sum << "\n";
      }
    }
  }
  else {
    // + recursive branch
    build(number, index + 1, runningSum + (number.at(index) - '0'), expression + "+" + number.at(index), number.at(index) - '0', "+", sum, length);
    // - recursive branch
    build(number, index + 1, runningSum - (number.at(index) - '0'), expression + "-" + number.at(index), number.at(index) - '0', "-", sum, length);
    // concatenation branch, must recalculate runningSum
    if (prevop.compare("+") == 0) { // subtract previous operand, add new operand
      build(number, index + 1, runningSum - prev + (stoi(to_string(prev) + number.at(index))), expression + "" + number.at(index), (prev * 10) + (number.at(index) - '0'), prevop, sum, length);
    }
    else if (prevop.compare("-") == 0) // add previous operand, subtract new operand
    {
      build(number, index + 1, runningSum + prev - (stoi(to_string(prev) + number.at(index))), expression + "" + number.at(index), (prev * 10) + (number.at(index) - '0'), prevop, sum, length);
    }
    else if (prevop.compare("") == 0) // continue concatenation
    {
      build(number, index + 1, stoi(to_string(prev) + number.at(index)), expression + "" + number.at(index), stoi(to_string(prev) + number.at(index)), prevop, sum, length);
    }
  }
} // end build

File: Assn5.py
# -*- coding: utf-8 -*-
import sys

number = sys.argv[1]
sum1 = int(sys.argv[2])
intNumberLength = len(number)
index = 0
runningSum = 0
expression = ""
prev = 0
prevop = ""

'''
* Create and evaluate expressions through recursive method.
   * number = original string
   * index = current index in number string
   * runningSum = evaluted sum of current expression
   * expression = string representation of current combination of operators+operands
   * prev = most recent operand
   * prevop = previous operator (+, -, or _/concatenation)
   * sum = target sum for comparison
   * length = length of number string
'''

def recursion(number,index,runningSum,expression,prev,prevop,sum1,length):
    if (index == length -1): # base case
        if (runningSum + int(number[index]) == sum1): # evaluate + last operand
            print(expression + "+" + str(number[index]) + "=" + str(sum1))
        if (runningSum - int(number[index]) == sum1): # evaluate - last operand
            print(expression + "-" + str(number[index]) + "=" + str(sum1))
		# evaluate operand concatenated with previous operand
        newEnd = int(str(prev) + str(number[index])) # concatenate previous operand with final digit
        if (prevop == "+" and (runningSum - prev + newEnd == sum1)): # if prevop = +, subtract previous operand and add new final operand
            print(expression + str(number[index]) + "=" + str(sum1))
        elif (prevop == "-" and (runningSum + prev - newEnd == sum1)): # if prevop = -, add previous operand and subtract new final operand
            print(expression+str(number[index]) + "=" + str(sum1))
        elif prevop == "": 
            expression += number[index]
            if int(expression) == sum1: # evaluate entire expression
                print(expression + "=" + str(sum1))
    else:
		# + recursive branch
        recursion(number, index +1, runningSum + int(number[index]), expression + "+" + str(number[index]),int(number[index]),"+",sum1,length)
		# - recursive branch
        recursion(number, index +1, runningSum - int(number[index]), expression + "-" + str(number[index]),int(number[index]),"-",sum1,length)
		# concatenation branch, must recalculate runningSum
        if prevop == "+": # subtract previous operand, add new operand
            recursion(number, index+1, runningSum - prev+ (int(str(prev) + number[index])),expression + ""+ number[index],(int(str(prev) + number[index])),prevop,sum1,length)
        elif prevop == "-": # add previous operand, subtract new operand
            recursion(number, index+1, runningSum+prev-(int(str(prev) + number[index])),expression + ""+number[index],(int(str(prev) + number[index])),prevop,sum1,length)
        elif prevop == "": # continue concatenation
            recursion(number,index+1, int(str(runningSum)+number[index]),expression +""+number[index],int(str(prev)+number[index]),prevop,sum1,length)

if(intNumberLength == 1): # prevents going into recursive loop
	if(int(number) == sum1):
		print(number + "=" + str(sum1))
else: # enter recursive loop - once with original string, another with negative string
    recursion(number,1,int(number[0]),number[0],int(number[0]),"",sum1,intNumberLength)
    recursion("-" + number, 2, int(number[0]) *-1, "-" + number[0], int(number[0])*-1,"",sum1,intNumberLength+1)



